hashes/src/sha256t.rs

```diff
@@ -112,26 +207,119 @@ fn from_engine<T: Tag>(e: sha256::HashEngine) -> Hash<T> {
+            /// The digest created by hashing `bytes` with engine's hashing algorithm.
+            #[allow(clippy::self_named_constructors)] // `hash` is a verb but `Hash` is a noun.
+            pub fn hash(bytes: &[u8]) -> Self {
+                let inner = $crate::sha256t::Hash::<$tag>::hash(bytes);
+                Self(inner)
+            }
+
+            /// Returns a hash engine that is ready to be used for data.
+            pub fn engine() -> $crate::sha256t::Engine<$tag> { $tag::engine() }
+
+            /// Creates a `Hash` from an `engine`.
+            ///
+            /// This is equivalent to calling `Hash::from_byte_array(engine.finalize())`.
+            pub fn from_engine(engine: $crate::sha256t::Engine<$tag>) -> Self {
+                let inner = $crate::sha256t::Hash::<$tag>::from_engine(engine);
+                Self(inner)
+            }
+
+            /// Zero cost conversion between a fixed length byte array shared reference and
+            /// a shared reference to this Hash type.
+            pub fn from_bytes_ref(bytes: &[u8; 32]) -> &Self {
+                // Safety: Sound because Self is #[repr(transparent)] containing [u8; Self::LEN]
+                unsafe { &*(bytes as *const _ as *const Self) }
+            }
+
+            /// Zero cost conversion between a fixed length byte array exclusive reference and
+            /// an exclusive reference to this Hash type.
+            pub fn from_bytes_mut(bytes: &mut [u8; 32]) -> &mut Self {
+                // Safety: Sound because Self is #[repr(transparent)] containing [u8; 32]
+                unsafe { &mut *(bytes as *mut _ as *mut Self) }
+            }
+
+            /// Copies a byte slice into a hash object.
+            pub fn from_slice(sl: &[u8]) -> Result<Self, $crate::FromSliceError> {
+                let inner = $crate::sha256t::Hash::<$tag>::from_slice(sl)?;
+                Ok(Self(inner))
+            }
+
+            /// Constructs a hash from the underlying byte array.
+            pub fn from_byte_array(bytes: [u8; 32]) -> Self {
+                let inner = $crate::sha256t::Hash::<$tag>::from_byte_array(bytes);
+                Self(inner)
+            }
+
+            /// Returns the underlying byte array.
+            pub fn to_byte_array(self) -> [u8; 32] { self.0.to_byte_array() }
+
+            /// Returns a reference to the underlying byte array.
+            pub fn as_byte_array(&self) -> &[u8; 32] { self.0.as_byte_array() }
+
+            /// Returns an all zero hash.
+            ///
+            /// An all zeros hash is a made up construct because there is not a known input that can
+            /// create it, however it is used in various places in Bitcoin e.g., the Bitcoin genesis
+            /// block's previous blockhash and the coinbase transaction's outpoint txid.
+            pub fn all_zeros() -> Self {
+                let inner = $crate::sha256t::Hash::<$tag>::all_zeros();
+                Self(inner)
+            }
```

Comment:
This one probably shouldn't be `pub`.

Also `const fn` for most functions.

Reply:
Re the `pub` thing, I don't follow the reasoning? Why should this one not be public but all the other types of hashes have this function public?

Re the const thing, answered elsewhere (MSRV limitation).

Reply:
It's the same as exposing hash engine publicly. Whether such hash makes sense depends on the context. E.g. it does in previous block hash//txid but doesn't when signing. Thus the newtype should be the one deciding if it's public.

But now I've realized that the only sane way to do that is to remove this completely and let the few newtypes that need `all_zeros` redefine it.